### PR TITLE
Qemu 7 QPP PR

### PR DESCRIPTION
**Pull request by lisi-w**

### CODE REVIEW

File: tests/plugin_qpp/qpp_client.c [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+QEMU_PLUGIN_EXPORT const char *qemu_plugin_name = "qpp_client";
+QEMU_PLUGIN_EXPORT int qemu_plugin_version = QEMU_PLUGIN_VERSION;
+QEMU_PLUGIN_EXPORT const char *qemu_plugin_uses[] = {"qpp_srv", NULL};
+
+#include "qpp_srv.h"
+
+void my_cb_exit_callback(gpointer evdata, gpointer udata);
+
+QEMU_PLUGIN_EXPORT void my_cb_exit_callback(gpointer evdata, gpointer udata) {
+    qemu_plugin_outs("called my on exit callback\n");
```
Comment (AndrewFasano) [inline tests/plugin_qpp/qpp_client.c:15]: Would there be a way to make this test fail if this doesn't get called? Perhaps some kind of global variable gets set in here and then on exit (but somehow after the on_exit cb?) we assert that it's set

File: include/qemu/qemu-plugin.h [hunk 2]
```diff
@@ -38,6 +39,11 @@
  */
 typedef uint64_t qemu_plugin_id_t;
 
+/**
```
[the remark this answers is not in the document]
Comment (lisi-w) [reply include/qemu/qemu-plugin.h:42]: unfortunately looks like it has to be in both -- qemu_plugin.h and plugin.h both contain functions that use it but are imported by different things

File: plugins/plugin.h [hunk 3]
```diff
@@ -47,13 +47,22 @@ struct qemu_plugin_state {
     struct qht dyn_cb_arr_ht;
 };
 
+typedef void (*cb_func_t) (gpointer evdata, gpointer udata);
+
+struct qemu_plugin_qpp_cb {
+    const char *name;
+    cb_func_t registered_cb_funcs[QEMU_PLUGIN_EV_MAX];
+    int counter;
```
[the remark this answers is not in the document]
Comment (lisi-w) [reply plugins/plugin.h:55]: yes, this is used to create, register, and run callbacks. I did realize I wasn't utilizing the counter though so I fixed that.

File: plugins/api.c [hunk 4]
```diff
@@ -400,6 +400,28 @@ bool qemu_plugin_bool_parse(const char *name, const char *value, bool *ret)
     return name && value && qapi_bool_parse(name, value, ret, NULL);
 }
 
+/*
+ * QPP: inter-plugin function resolution and callbacks
+ */
+
+gpointer qemu_plugin_import_function(const char *target_plugin,
```
[the remark this answers is not in the document]
Comment (lisi-w) [reply plugins/api.c:407]: I tried to make sure each separate commit built, and the plugin-qpp file is required for this to build

File: tests/plugin/bb.c [hunk 5]
```diff
@@ -15,6 +15,7 @@
 #include <qemu-plugin.h>
 
 QEMU_PLUGIN_EXPORT int qemu_plugin_version = QEMU_PLUGIN_VERSION;
+QEMU_PLUGIN_EXPORT const char *qemu_plugin_name = "bb";
```
[the remark this answers is not in the document]
Comment (lisi-w) [reply tests/plugin/bb.c:18]: fixed

File: tests/tcg/arm/Makefile.softmmu-target [hunk 6]
```diff
@@ -24,3 +24,4 @@ test-armv6m-undef: EXTRA_CFLAGS+=-mcpu=cortex-m0 -mfloat-abi=soft
 
 run-test-armv6m-undef: QEMU_OPTS+=-semihosting -M microbit -kernel
 run-plugin-test-armv6m-undef-%: QEMU_OPTS+=-semihosting -M microbit -kernel
+run-plugin-qpp-test-armv6m-undef: QEMU_OPTS+=-semihosting -M microbit -kernel
```
[the remark this answers is not in the document]
Comment (lisi-w) [reply tests/tcg/arm/Makefile.softmmu-target:27]: yes, make check-tcg fails without it

File: include/qemu/plugin-qpp.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+ * dynamically resolve a target function in another plugin or raise a fatal
+ * error on failure. In particular, it must handle the following two cases:
+ * 1) When the header is loaded by the plugin that defines the function.
+ *    In this case, we do not need to find the symbol externally.
+ *    qemu_plugin_name_to_handle will return NULL, we see that the
+ *    target plugin matches qemu_plugin_name (required to be defined by
+ *    plugins of API v2+) and do nothing.
+ * 2) When the header is loaded by another plugin. In this case
+ *    we get the function pointer from qemu_plugin_import_function
+ *    and correctly cast and assign the function pointer
```
[the remark this answers is not in the document]
Comment (lisi-w) [reply include/qemu/plugin-qpp.h]: updated this, thanks!

File: tests/plugin_qpp/qpp_client.c (hunk 1 above)
Comment (lisi-w) [reply tests/plugin_qpp/qpp_client.c:15]: I can't think of a way to assert after the cb as it should be the last thing to run

**Pull request closed without merging**